[RLlib; Offline Data] Give users option to customize `OfflineData`.

simonsays1980 opened the pull request:
## Why are these changes needed?

So far, users can use either a custom connector for the learner connector pipeline in the `OfflinePreLearner` or override the `OfflinePreLearner` to modify the offline data pipeline. In some cases this could be not enough (e.g. when data is in encoded formats not directly readable by `ray.data`). 
This PR proposes a change to give users the option to customize also the complete data pipeline by overriding the `OfflineData` class and define therein how data should be read and transformed prior to the `OfflinePreLearner`. It could however also be used to define a simple pipeline not relying on Ray Data at all, but instead making use of other data pipeline framework (e.g. `datasets` from HuggingFace, or `h5py`).

It does so by
- Providing an `offline_data_class` attribute in the `AlgorithmConfig` that could be provided in the `offline_data` method by users.
- Using the custom class or RLlib's default `ray.rllib.offline.offline_data.OfflineData`.

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

simonsays1980 committed:
Added keyword arguments to 'OfflinePreLearner.__init_' and 'OfflinePreLearner._map_to_episodes' methods.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Added customization option for 'OfflineData' classes and wrote a test for it. In addition added keyword arguments to the 'OfflinePreLearner' mmethods such that overriding offers more customization.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Added docstring for 'offline_data_class' in ÄALgorithmConfig.offline_data'.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Merge branch 'master' into offline-rl-make-offline-data-overrideable

sven1977 committed:
[RLlib] APPO (new API stack) enhancements vol 05: Auto-sleep time AND thread-safety for MetricsLogger. (#48868)

sven1977 commented on rllib/algorithms/algorithm.py:
```diff
@@ -710,7 +710,10 @@ def setup(self, config: AlgorithmConfig) -> None:
             from ray.rllib.offline.offline_data import OfflineData
 
-            self.offline_data = OfflineData(self.config)
+            # Use either user-provided `OfflineData` class or RLlib's default.
```
Awesome improvement!

sven1977 commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -2602,6 +2604,14 @@ def offline_data(
+                data or your loss. Usually overriding the `OfflinePreLearner` and using
+                the resulting customization via `prelearner_class` suffices for most
+                cases. The default is `None` which uses the base `OfflineData` defined
+                in `ray.rllib.offline.offline_data`.
```
```suggestion
                in `ray.rllib.offline.offline_data.OfflineData`.
```

sven1977 commented on rllib/algorithms/algorithm_config.py:
```diff
@@ -2782,7 +2792,9 @@ def offline_data(
         """
         if input_ is not NotProvided:
-            self.input_ = input_
+            self.input_ = input_  #
```
```suggestion
            self.input_ = input_
```

sven1977 commented on rllib/offline/offline_data.py:
```diff
@@ -223,7 +223,7 @@ def sample(
     def default_read_method_kwargs(self):
         return {
-            "override_num_blocks": max(self.config.num_learners * 2, 2),
+            # "override_num_blocks": max(self.config.num_learners * 2, 2),
```
not sure I understand this change ...

simonsays1980 replied:
It shouldn't be there. Was done during testing. In the streaming optimization PR this will be removed from the default arguments and all examples.

simonsays1980 replied:
Great catch btw! :)

sven1977 reviewed (approved):
Looks great! Just one question on the commented out config line (see my comments). Thanks @simonsays1980

sven1977 committed:
Apply suggestions from code review

Signed-off-by: Sven Mika <sven@anyscale.io>

Jay-ju committed:
[data] provide more messages when webdataset format is error (#48643)

<!-- Thank you for your contribution! Please review
https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before
opening a pull request. -->
When the dataset of webdataset consists of many tar files, if there is a
failure in reading one of the files, it is necessary to print the name
of the failed file.

before this pr:
```
ValueError: fx001.jpg: duplicate file name in tar file jpg dict_keys(['__key__', 'jpg', 'json'])
```

after this pr:

```
ValueError: fx001.jpg: duplicate file name in tar file jpg dict_keys(['__key__', 'jpg', 'json']), in emr-autotest/datasets/jkj_10/1b63122e734064926c47a2ccdcbaafd3.tar
```



<!-- Please add a reviewer to the assignee section when you create a PR.
If you don't have the access to it, we will shortly find a reviewer and
assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this
solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git
commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for
https://docs.ray.io/en/master/.
- [ ] I've added any new APIs to the API Reference. For example, if I
added a
method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a
few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

Signed-off-by: jukejian <jukejian@bytedance.com>

jjyao committed:
[Core] Retryable grpc client (#47981)

Signed-off-by: Jiajun Yao <jeromeyjj@gmail.com>

bveeramani committed:
[Data] Add writing benchmark (#49006)

We don't have a release test that tracks the performance of writing
data. This PR adds one.

---------

Signed-off-by: Balaji Veeramani <bveeramani@berkeley.edu>

sven1977 committed:
[RLlib] External Env support for new API stack: New example script and custom tcp-capable EnvRunner. (#49033)

simonsays1980 committed:
Remvoed debugging changes.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Merged master and resolved conflict in PR.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>

simonsays1980 committed:
Merge branch 'master' into offline-rl-make-offline-data-overrideable

simonsays1980 committed:
Merged master and resolved conflict.

Signed-off-by: simonsays1980 <simon.zehnder@gmail.com>
